### DISCUSSION TITLE
rendering a vsg in a Qt widget?

### DISCUSSION THREAD

**Discussion by brouhaha** [vsg-dev/VulkanSceneGraph#1189] [General]
Is there any way to render a vsg into a Qt Widget? I was glad to see that there is Qt support provided as vsgQt, but I'd like to render the scene graph in a widget because I want to have a menu bar and various other widgets in the same window.

I don't want the user to see an MDI, but if there's no other way to display a vsg in a widget, is it possible to put it in an MDI area in a single MDI window that is always maximized (within the MDI area) and has no window decorations (title bar, etc.)?

Thanks!

**Comment by robertosfield** [comment]
I'm no Qt expert, I learned enough Qt to implement vsgQt, so I'll have to defer to others with more expertise on how best to tackle more niche topics.

It not clear to me by what you mean by render VSG in Qt Widget as this is so open ended.  Could you provide a screenshot illustration of what you mean?

**Comment by rainergericke** [comment]
I think brouhaha is meaning something like this:
<img width="500" alt="SoQtExaminerViewer" src="https://github.com/vsg-dev/VulkanSceneGraph/assets/31243285/693fd08d-94b4-491a-836b-ebae2c656449">
It is not a vsg widget, but a Qt6 based program which displays a cone generated by Coin3d. Maybe it can be used as a starting point.

**Reply by robertosfield** [reply]
Um... this type of stuff is so open ended and application specific I'd be wary of attempting to include stuff like this is vsgQt. 

Folks are welcome to write their own and publish it as a library or a PR for vsgQt.

**Reply by brouhaha** [reply]
rainergericke is correct that that's the sort of thing I want to do. I want to be able to put one or more vsg renderings into widgets that are inside a "normal" Qt window along with other "normal" Qt Widgets.
As opposed to the vsg renderings being in their own top-level windows.

**Comment by VasanthBalguri** [comment]
I don't think it is possible out of box (as of version 6, there is no QVulkanWidget present), will need to make a completely custom widget in qt.
This might be a big undertaking which might not be portable across different versions of Qt.

somebody did implement QvulkanWidget though:
https://www.youtube.com/watch?app=desktop&v=2E_hc6CfmbA

**Comment by rainergericke** [comment]
I had a closer look into the Qt-6.7.1 example projects. There is no QVulkanWidget indeed, but a QVulkanWindow instance can be also used as a Widget. So it would be ok to write a class like QVsgWindow (well, somebody must do it). Her is a shot of one of the QtVulkanWindow examples:
<img width="1022" alt="QtVulkanExample" src="https://github.com/vsg-dev/VulkanSceneGraph/assets/31243285/b195acaa-c251-479f-9725-71180aa0ef28">

**Reply by brouhaha** [reply]
Thanks! I didn't realize that a QVulkanWindow can be used as a QWidget. That definitely could be helpful in moving toward my objective.

**Comment by VasanthBalguri** [comment]
Think thats how vsgQt works if I am not wrong? Am I missing something?

**Reply by rainergericke** [reply]
No, not exactly. One difference is, that vsgQt does not work an macOS due to weird window pointer issues. The above example works on the Mac. Maybe we can take some of the vsgQt ideas for a QtVsgWindow class.

**Comment by VasanthBalguri** [comment]
Okay, I can see some issue with my wayland implementation as well.
Please share the link if possible or how to find it online.

**Reply by rainergericke** [reply]
The Qt Vulkan examples are contained in the Qt Open Source distribution. The fastest way is to load the installer program from [https://www.qt.io](url). You can build the examples easily with the Qt Creator.

**Comment by robertosfield** [comment]
One issue to be aware of with Qt's Vulkan integration ties the vkDevice per Window (and I'd guess Widget) which is a bit like OpenGL's associated of the window with a graphics context.  

However, this approach prevents sharing a vk/vsg::Device between Windows which is really inefficient for multiple window applications.  Far better is to have a vsg::Device per GPU and have all the windows shared as then you can share all the Vulkna objecrts and GPU memory between Windosws.  It for this reason I abandoned the use of Qt's Vulkan integration and implemented the Vulkan support in vsgQt using the VSG's native vkSurface support.

I don't know if it's possible to create a general purpose Qt Widget which can have a vkSurface created for it, but perhaps the Qt examples might help suggest a way.  If not then perhaps one could render to a colour buffer and then copy this to the widget without doing a round trip to the GPU to CPU and back to the GPU.

**Reply by VasanthBalguri** [reply]
>  If not then perhaps one could render to a colour buffer and then copy this to the widget without doing a round trip to the GPU to CPU and back to the GPU.

This seems like a good hack, but I am not sure of its overall feasibility or what the implications can be for its portability....

**Comment by nolankramer** [comment]
Hi all - please see https://github.com/vsg-dev/vsgQt/issues/39#issuecomment-2323536579 for a possible implementation.

As a note - I'll be creating an open-source MIT licensed repo that implements that methodology.

@robertosfield - if the experiment/repo proves successful, is there a chance to hold the project upstream under vsg-dev? It would greatly aid in visibility for users.

**Comment by robertosfield** [comment]
Currently my inclination is that these solutions would be start with a dedicated vsgQt6 library and if it looks appropriate for inclusion with vsg-dev then I'm open to this.  Having this projects be maintained long term is important, I don't have the time and skills to support too many disparate projects if original developers move on.

**Comment by jacky-wang7** [comment]
Hi All,
I found a simple way to rendering a vsg in a Qt widget. 
Using QMdiArea, like this:
```
vsgQt::Window* renderWindow = vsgQt::Window(); 
// The window container has a number of known limitations.
QWidget* renderWidget =  QWidget::createWindowContainer(renderWindow);

auto widget = new QWidget(parent);
auto mdiArea = new QMdiArea(widget);
 // Hide the status bar of mdiArea
mdiArea->addSubWIndow(widget, Qt::CustomizeWindowHint | Qt::FramelessWindowHint | Qt::WindowTitleHint);
mdiArea->showMaximized();
renderWidget->showMaximized();

//Now use the widget like a normal QT Widget
widget->lower();
```
Disadvantages of container: [createWindowContainer](https://doc.qt.io/qt-6/qwidget.html#createWindowContainer)
At present, everything works fine when I use this method, but I think this should not be the most correct method. I am not an expert in Qt, and this is relatively simple.
